source/material_model/rheology/composite_visco_plastic.cc

```diff
@@ -0,0 +1,372 @@
+
+        if (use_dislocation_creep)
+          creep_edot_and_deriv = creep_edot_and_deriv + dislocation_creep->compute_strain_rate_and_derivative(creep_stress, pressure, temperature, dislocation_creep_parameters);
+
+        if (use_peierls_creep)
+          creep_edot_and_deriv = creep_edot_and_deriv + peierls_creep->compute_strain_rate_and_derivative(creep_stress, pressure, temperature, peierls_creep_parameters);
+
+        const double strain_rate = creep_stress/(2.*max_viscosity) + (max_viscosity/(max_viscosity - min_viscosity))*creep_edot_and_deriv.first;
+        const double strain_rate_deriv = 1./(2.*max_viscosity) + (max_viscosity/(max_viscosity - min_viscosity))*creep_edot_and_deriv.second;
+
+        return std::make_pair(strain_rate, strain_rate_deriv);
+      }
+
+
+
+      template <int dim>
+      void
+      CompositeViscoPlastic<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.declare_entry ("Include diffusion creep", "true",
+                           Patterns::Bool (),
+                           "Whether to include diffusion creep in the rheological formulation.");
+
+        prm.declare_entry ("Include dislocation creep", "true",
+                           Patterns::Bool (),
+                           "Whether to include dislocation creep in the rheological formulation.");
+
+        prm.declare_entry ("Include Peierls creep", "true",
+                           Patterns::Bool (),
+                           "Whether to include Peierls creep in the rheological formulation.");
+
+        // Diffusion creep parameters
+        Rheology::DiffusionCreep<dim>::declare_parameters(prm);
+
+        // Dislocation creep parameters
+        Rheology::DislocationCreep<dim>::declare_parameters(prm);
+
+        // Dislocation creep parameters
+        Rheology::PeierlsCreep<dim>::declare_parameters(prm);
+
+        // Some of the parameters below are shared with the rheology models,
+        // and will already have been declared. This is fine, the deal.II
+        // parameter handler allows repeated declarations. The default values
+        // below will override any defaults given in the rheology models.
+        // The new defaults will apply to both the material model and the
+        // underlying rheology modules.
+        prm.declare_entry ("Minimum strain rate", "1.4e-20", Patterns::Double(0.),
+                           "Stabilizes strain dependent viscosity. Units: \\si{\\per\\second}.");
+
+        // Viscosity iteration parameters
+        prm.declare_entry ("Strain rate residual tolerance", "1e-22", Patterns::Double(0.),
+                           "Tolerance for correct diffusion/dislocation strain rate ratio.");
+        prm.declare_entry ("Maximum creep strain rate iterations", "40", Patterns::Integer(0),
+                           "Maximum number of iterations to find the correct "
+                           "creep strain rate.");
+
+        // Strain rate and stress limiting parameters
```

Comment:
Another questions about the minimum and maximum viscosities. Here, there is a unique value for each compositional field, whereas in visco plastic and diffusion dislocation they are single values. Could this cause any unforeseen issues?

Reply:
I don't think so (or at least, if there are, they are unforeseen for both of us!). Single values are also accepted for multiple compositional fields.

As an aside, I don't think diffusion-dislocation will end up using this composite, because of the hard limited viscosity in that model. We can discuss this and other minor issues when we start *the big tidy* that @gassmoeller dreams about :)